### ISSUE TITLE
Reordering tasks

### ISSUE THREAD

**Issue by doranism** [mpcjanssen/simpletask-android#636]
As I wrote in #458, I would like to make the following suggestion to improve 
1. sorting tasks in the "opportunity-now"-list via threshold date to adjust the order I want to work on them.
2. planning the tasks for the upcoming week including the possibility to see the amount of work at the day, where the task is pushed. 

This is important, because the priority of a task is often relative and depends on how many tasks have to be completed on the same day.

To achieve that, my proposal aims to set the values (dates, prios, tags...) relative to other tasks in the current view and perform many small changes to rearrange the tasks. The reorder-action is based on the grouping-criterion and the 1. sorting criterion of the current view. The user will be able to change the values of the selected task in the following way:

![reorder](https://cloud.githubusercontent.com/assets/22551922/19091311/671224fe-8a82-11e6-9350-9481127f1b61.png)

**1. Select a task.** Reorder-icon in top-bar shows up.
**2. Activate reorder-mode** by tapping the icon. Reorder-icon is highlighted. All groups collapse to their headline, except for the current. Headlines of the groups show the number of the tasks contained.
**3. Scroll the list up or down**. Selected task is fixed to the current position, list scrolls underneath the selected task. The changed value is shown in the properties of the task.
**4. Stop scrolling**. Current value is selected. If the task has been moved to another group: open it.
Optional: scroll again (go to 3.) 
**5. Deselect**. Closed groups open up again. 
Optional: select another task, then go to 3. 
**6. Leave sort-mode** by tapping highlighted icon. Reset to the initial view.

To give the possibility of choosing a date before or after other tasks, there should show up groups +1 and -1 days of an existing date.

It is clear that many questions remain (editing more than one task, behavior in filtered views etc.). If you like the concept, I would be happy to discuss the details with you!

P.S: Please take this propasal as a contribution of thoughts and feel free to do what ever you want with it :-)

**Comment by smichel17** [comment]
Alright, I finally got a chance to look at this.

I agree with your assessment of simpletask's (lack of) features in these regards. I also think that your suggested would do a good job of adding those features. However, I don't think we should implement it as-is:

- I'm not really sure how it would work when you have many tasks selected, especially when they are not all on screen at once.
- When you exit this mode, the file will need to have changed in some way so that the tasks will be sorted according to what you just set. It's not particularly clear how we'll do that.
- That's an old screenshot; on simpletask right now, there are many more actions in the app bar during selection; I don't think we have room for another on smaller devices.
- This is reasonably intuitive, but I think we can do better.
- I am concerned about simpletask becoming too much like a calendar app -- I think we should be showing you what tasks are available to do, but we should not be storing information about when you plan on doing them. That's why we have the "create calendar appointment" function.

Instead, I think we could address your use cases with two separate features:

1. Long press and drag to re-order.
2. Improvements on #603 -- that way you can set a ca
2a. Move "Create calendar appointment" to be on a floating action button during selection; it should be more advertised.

(1) Allows you to rearrange your tasks easily. (2) Allows you to create calendar appointments and then also to view your calendar alongside your tasks, so you can see what you've scheduled to do that day and also what else is available to do.

**Comment by doranism** [comment]
Today I got the Update from f-droid to V8.3.0. It really is a big improvement! Especially setting due or threshold dates, which I do alot.

> how it would work when you have many tasks selected

I thought about that but tried to explain the basic case first. As you said, the easier solution would be, to only allow reordering one task at once. But it can be useful to reorder more than one task. Maybe it could be done like this:
1. Select tasks
2. Activate reorder mode or drag tasks: selected tasks are bundled to one item (the one, where the user started to drag) with the title "X selected tasks".
The bundle of tasks could be handled like a single task, which means that all included tasks will get the same destination-value. As you wrote in #406, it should always be changed the value of the source.
This approach will also work, if the selected tasks are not shown on a screen at once.

> the file will need to have changed in some way so that the tasks will be sorted according to what you just set

Can you please explain this a little bit more what you mean?

> there are many more actions in the app bar during selection

I saw that, after I updated Simpletask and it's true, there is not enough space left. In addition to that, I think that in my first suggestion too many klicks are necessary to reorder a task. Maybe it would be a better solution, if selected tasks show up a handle to drag up or down. In this way it isn't necessary to activate and deactivate an reorder-mode.

> 1. Long press and drag to re-order

Yes that's another good option. I did not suggest a long press, because it is already used for opening links contained in a task. IMO Dragging the selected task instead of moving the list underneath has the disadvantage that, if you reach the edge of the screen, scrolling gets slow and you don't see what comes next.

> "Create calendar appointment"

To me it is important that I can easily edit the date of a task. Therefore your first suggestion would better fit my needs. I don't plan the exact time of my tasks, it's still a list sorted by threshold. I only estimate, if I can manage to complete the amount of tasks I assigned to a day.

**Comment by smichel17** [comment]
> Today I got the Update from f-droid to V8.3.0. It really is a big improvement! Especially setting due or threshold dates, which I do alot.

I agree :)

>As you wrote in #406, 

I'd forgotten about #406. :P

> selected tasks are bundled to one item

We'd need to add a cancel button, in case I start to scroll them, then decide I actually want to cancel the movement. That shouldn't be too hard, though, since we can remove options like archive and delete from the app bar when scrolling.

This is the kind of thing I'd want to actually use for a while to see whether my actual experience lines up with my analysis, but overall it seems reasonable.

>>the file will need to have changed in some way so that the tasks will be sorted according to what you just set

>Can you please explain a little bit more what you mean?

Take the example in #406, but what if it were like this? Same example as the last one in #406, but we're dragging `a` into `@c` instead of the other way around. Starting list:
```
a @a @b
b @b
c @c
```
I drag `a` into `@c`. Result?
```
a @c
b @b
c @c
```
or
```
a @a @b @c
b @b
c @c
```
or
```
a @b @c
b @b
c @c
```

In #406, I said that option 3 makes the most sense. But in this example, option 3 would leave `a` displayed under the `@b` header instead of under `@c`, where we dragged it. So, like #406, we need #153 before we can do this. And like #153, we need #617 first.

Then we need to figure out how to handle threshold/due/creation date, priority, file order, and alphabetical. Some of those are easy and some are hard.

Then we figure out what happens if I drag this task down below another task inside the same header? Does it have no effect? Does it change based on whatever secondary sort is enabled? Actually, the same question stands if I drag to a different header. Do secondary attributes also change or only the primary one? If the latter, why not just use the existing functions for changing lists/tags/prio/etc? (For the record, I think we can make the existing functionality slightly faster & more convenient than it currently is).

What if I'm using a custom onGroup function? Maybe if onGroup is defined, we also must require the user to define an "onDrag" function that takes a task, where it was dragged from and to, and returns a modified task. Or just keep drag disabled in that case. Are either of those tradeoffs we're willing to make? What if we decide to add a lua callback for sorting?

My point is that due to the nature of todo.txt, we can't just say "show this task below that one"; there has to be a reason for that in the text file, and that gets very complex very quickly. To be clear, this is also an issue for single-task drag, which is therefore a low priority. That doesn't mean I relish adding the additional variable of multi-task drag on top of that... let's figure out one at a time first, then go from there.

>I think that in my first suggestion too many klicks are necessary to reorder a task. Maybe it would be a better solution, if selected tasks show up a handle to drag up or down. In this way it isn't necessary to activate and deactivate an reorder-mode.

In that case you give up scrolling the list and it becomes harder to make an intuitive interface for rearranging multiple tasks.

>Long press and drag [... is] another good option. I did not suggest a long press, because it is already used for opening links contained in a task. IMO Dragging the selected task instead of moving the list underneath has the disadvantage that, if you reach the edge of the screen, scrolling gets slow and you don't see what comes next.

Dragging and scrolling offer two things that the current system does not: they are very visual, and they allow you to quickly change multiple attributes with one action (either to get through an inbox or to sort tasks the way you want without worrying about exactly what attributes need to be changed).

Drag has some advantage over scroll in that it's more tactile and is only one gesture, that I can perform without selecting a task first. And it has the disadvantage that you pointed out.

>To me it is important that I can easily edit the date of a task.

Which date (creation, due, threshold)?

>Therefore your first suggestion would better fit my needs. I don't plan the exact time of my tasks, it's still a list sorted by threshold. I only estimate, if I can manage to complete the amount of tasks I assigned to a day.

What if tasks could appear under multiple headers, and you could define a function so that tasks would only appear under day X's header if the threshold date is before day X *and* the due date is after it?

**Comment by doranism** [comment]
Now we're really getting into it :-)

I wish that I could reply faster! But I have alot to do at the moment.

> We'd need to add a cancel button

Yes, that's an good idea

> This is the kind of thing I'd want to actually use for a while to see whether my actual experience lines up with my analysis

I think that the whole function will need some testing on the usability. It only makes sense, if it is more fluent than the way attributes can be set right now. But in my opinion it is worth trying.


> example in #406

In general, the conclusion of my thoughts about reordering tasks is, that it should
a) be based very much on what is currently visible on the screen and 
b) change the values relative to other tasks,
because otherwise it doubles up the existing functionality.

In your example I would expect exactly the result that you described and I agree that #153 will be necessary to achieve that. By the way, #153 is the reason why I assign tasks only to one list and I almost never sort after tags, because the result is unclear to me.


> how to handle threshold/due/creation date, priority, file order, and alphabetical

There are four types of data:

1. The values of lists and tags follow continously one after another and each list or tag is used at least by one task. There are no gaps in between their values. So dragging a task to e.g. another list should delete the list of the source and set the list of the destination. Afterwards the task should be sorted in > the way the current view is set up.

2. Instead, there are many possible values for dates and priority but only few of them will be used in the tasklist. This means that there can be gaps in between or at the end. 
If I drag a task to another date, it should work like 1. To give the possibility to drag a task to a date before or after an existing date, there should show up additional headers or slots. For example, in a tasklist the threshold-dates 4.3. 11.3. and 13.3. are used. So by enabling reorder-mode, the headers 
3.3. **4.3.** 5.3. and 10.3. **11.3.** 12.3. **13.3.** 14.3. will be shown. The additional dates could also be shown only graphical before and after an existing date.

3. Fileorder is either very easy or very difficult, depending on the task-numbers in the file. If the the task-numbers have to be continously counting up in the file, many of them have to be changed on reorder. In this case I would suggest to not allow reordering. If they don't, it would be like a manual reorder.

4. Changing the alphabetical order does only make sense to me, if one uses prefixes at the beginning of the text of the tasks. Normal text would change something like this: from "Add sugar..."  to "Bdd sugar..."


> what happens if I drag this task down below another task inside the same header?

Starting again from what is displayed to the user, the task will be shown at a certain position in a list under a certain header. As I suggested in my first posting, at first the task should be dragged through the list under the current header, which means, that the secondary attribute will be changed.
If the user drags the task to another header, the secondary attribute will be set back to the original value and the primary attribute will start to change. 
To separate between changing the secondary and the primary attribute, I suggested to close all groups, exept for the current, when reorder-mode is enabled. This accellerates scrolling or dragging through the headers aswell.
On stop, the first attribute will be selected according to the current header and the current group will be opened. Now the user can drag again in the list of the new group to change the secondary attribute.

I also thought about other possibilities like scrolling up or down for to change the position inside of a group (secondary attribute) and swiping left or right to skip to the next group (secondary attribute). But because Mark Janssen wrote somewhere that he does not like swiping gestures because of the confusion with opening the drawers, I stayed to the suggested version. Maybe we can disable opening the drawers during reorder-mode or we can think of other better solutions.


> What if I'm using a custom onGroup function?

I do not know very much about Lua-scripting, but I understood that it is possible to use a range for grouping (e.g. before today / today / after today). In these cases it is very difficult to set a value in a way, that the user understands. Maybe we could use today-1 for date < today in the onGroup function. Alternatively we could restrict dragging to the current list (secondary value) in these special cases.


> that gets very complex very quickly

yes that is true, but its "only" complex in the background and to me it seems that the problems can be solved. I expect it to be a big improvement in usability though.


> In this way it isn't necessary to activate and deactivate an reorder-mode. 
> In that case you give up scrolling the list

I meant that the reorder-mode will be activated when dragging is started and it ends, after deselecting. Before and after that it would be possible to scroll and select tasks.


> Which date?

Mostly threshold, sometimes due, never creation.
